coral-schema/src/main/java/com/linkedin/coral/schema/avro/SchemaUtilities.java

```diff
@@ -436,48 +439,74 @@ private static boolean isUnionSchemaCompatible(@Nonnull Schema leftSchema, @Nonn
       case INT:
       case LONG:
       case STRING:
-      case NULL:
-        return leftSchema.getType() == rightSchema.getType();
+        return leftSchema.getType() == rightSchema.getType() ? leftSchema : null;
```

Comment:
Do we have to keep rechecking `leftSchema.getType() == rightSchema.getType()` or extract it to the top before the switch case (with the exception of null and nullable union which can be handled as special case)?
Also I think pulling up NULL and nullable union checks before the switch case could simplify the code. 
Also, can we somehow unify the code for NULL and nullable union? Seems both will result in adding null to the result.

Reply:
I don't think it is very easy to extract and unify, which might bring unexpected exceptions or cases we might fail to consider. And as there are still so many special cases to consider, the unified code for `union` and `null` will not be simpler than current code. I think to deal with `union` and `null` with other types together in `switch` is clearer and safer.

Besides, if we don't include `null` and `union` in `switch`, we need to add more code to determine the type for each side.
For example, for just one special case: `null` union `[null, int]`, we need to use:
```
if (leftSchema.getType() == Schema.Type.NULL && rightSchema.getType() == Schema.Type.UNION && AvroSerdeUtils.isNullableType(rightSchema) || rightSchema.getType() == Schema.Type.NULL && leftSchema.getType() == Schema.Type.UNION && AvroSerdeUtils.isNullableType(leftSchema)) {
     return leftSchema.getType() == Schema.Type.UNION ? leftSchema : rightSchema;
}
```
We can extract the judgement as other functions, but I don't think it will be simpler than current patch.
@funcheetah What is your idea?